src/Microsoft.SqlTools.ServiceLayer/FileBrowser/FileBrowserService.cs

```diff
@@ -0,0 +1,264 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Data.Common;
+using System.Data.SqlClient;
+using System.Threading.Tasks;
+using Microsoft.SqlTools.Hosting.Protocol;
+using Microsoft.SqlTools.ServiceLayer.Connection;
+using Microsoft.SqlTools.ServiceLayer.Connection.ReliableConnection;
+using Microsoft.SqlTools.ServiceLayer.FileBrowser.Contracts;
+using Microsoft.SqlTools.ServiceLayer.Hosting;
+
+namespace Microsoft.SqlTools.ServiceLayer.FileBrowser
+{
+    /// <summary>
+    /// Main class for file browser service
+    /// </summary>
+    public sealed class FileBrowserService
+    {
+        private static readonly Lazy<FileBrowserService> LazyInstance = new Lazy<FileBrowserService>(() => new FileBrowserService());
+        public static FileBrowserService Instance => LazyInstance.Value;
+
+        // Cache file browser operations for expanding node request
+        private Dictionary<string, FileBrowserOperation> ownerToFileBrowserMap = new Dictionary<string, FileBrowserOperation>();
+        private ConnectionService connectionService = null;
+
+        private Dictionary<string, ValidatePathsCallback> validatePathsCallbackMap { get; set; }
+
+        /// <summary>
+        /// Signature for callback method that validates the selected file paths
+        /// </summary>
+        /// <param name="eventArgs"></param>
+        public delegate bool ValidatePathsCallback(FileBrowserValidateEventArgs eventArgs, out string errorMessage);
+
+        internal ConnectionService ConnectionServiceInstance
+        {
+            get
+            {
+                if (connectionService == null)
+                {
+                    connectionService = ConnectionService.Instance;
+                }
+                return connectionService;
+            }
+            set
+            {
+                connectionService = value;
+            }
+        }
+
+        /// <summary>
+        /// Service host object for sending/receiving requests/events.
+        /// Internal for testing purposes.
+        /// </summary>
+        internal IProtocolEndpoint ServiceHost
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        public FileBrowserService()
+        {
+            this.validatePathsCallbackMap = new Dictionary<string, ValidatePathsCallback>();
+        }
+
+        /// <summary>
+        /// Register validate path callback
+        /// </summary>
+        /// <param name="service"></param>
+        /// <param name="callback"></param>
+        public void RegisterValidatePathsCallback(string service, ValidatePathsCallback callback)
```

Comment:
>RegisterValidatePathsCallback [](start = 20, length = 29)

why isn't this done during construction, or during InitializeService?

Reply:
? this is called from InitializeService of DisasterRecoveryService